README.md
UniMux [![Build Status](https://travis-ci.org/carosio/unimux.svg)](https://travis-ci.org/carosio/unimux) [![Coverage Status](https://coveralls.io/repos/carosio/unimux/badge.svg?branch=master&service=github)](https://coveralls.io/github/carosio/unimux?branch=master)
======


UniMux applciation for routing [hello](https://github.com/travelping/hello)-based requests.

Minimal requirements are:

* erlang >= 17 (better 17.5)
* elixir >= 1.0 (1.1.0-dev)
* Apple Bonjour or a compatible API such as Avahi with it's compatibility layer along with the appropriate development files:
  * OS X - bundled
  * Windows - Bonjour SDK
  * BSD/Linux - search for Avahi in your operating systems software manager

If you install erlang on Ubuntu, install aditionally:

* erlang-parsetools
* erlang-eunit

Build an application:

    $> mix do deps.get, compile

Start an interactive shell with application:

    $> iex -S mix

Dev configuration is placed on `config/config.exs`

Release configuration is placed here: `unimux.conf`, but should not changed manuelly, instead change the schema: `unimux.schema.exs` and regenerate the configuration with:

    $> mix conform.configure

Build an test release locally:

    $> MIX_ENV=prod mix release

See more information about releases here: [exrm](https://github.com/bitwalker/exrm) and [relx](https://github.com/erlware/relx)

Build release on builder, which does download dependencies for you, use release enviroment, which overwrites the pathes to `deps/<app>`

    $> MIX_ENV=release mix release

Building the documentation:

    $> mix docs


CHANGELOG.md
# 1.3.3 (2016-11-16)

* Bug fixes
    * exrm and relx are updated to fix issue with non-readable sys.config

# 1.3.2 (2016-11-08)

* Bug fixes
    * various dependency updates to fix metrics configuration
    * `mock` is now a test dependency

# 1.3.1 (2016-10-28)

* Bug fixes
    * lager updated to 3.2.2 which contains bug fixes for log rotation
    * hello updated to 3.3.0 which ctaonins bug fix for requests timeouts
    * added missed exlager to conform schema

# 1.3.0 (2016-10-18)

* Enhancements
  * added support of elixir-1.3
  * unimux moved to conform and exrm from upstream

# 1.2.11 (2016-10-06)

* Bug fixes
  * Set listener IP for distributed erlang to localhost

# 1.2.10

* Enhancements

  * added support of elixir-1.2

# 1.2.9

* Enhancments
  * updated exrm and metricman (including influxdb reporter)

# 1.2.8

* Enhancements
  * update config files

# 1.2.7

* Enhancements
  * add default timeout for routes

# 1.2.6

* Enhancements
  * improve config translations
  * add configurable options for each route
  * update hello for improve logging

# 1.2.5

* Enhancements
  * Get rid of dnssd
  * update dependencies hello, metricman to have better logging and disable default metric backend

# 0.2.3

* Enhancements
  * dependency bump exrm and conform

# 0.2.2

* Enhancements
  * take exrm from xerions repository
  * add travis tests with Elixir 1.1.0 and Erlang 17.4
  * test conform schema
  * fix feature to exclude dependencies in release phase

# 0.2.1

* Enhancements
  * update hello version to 3.1.0
  * add travis
  * add metricman
  * add coverex

# 0.2.0

* Enhancements
  * update to hello with better http-client

* Backwards incompatible changes
  * a.b.c matched on a now, before it has not matched on it

# 0.1.4

* Bug fixes
  * fix .schema file

# 0.1.3

* Bug fixes
  * update jsx format dependency

# 0.1.2

* Bug fixes
  * update hello, to support http ip:port in clients without dns

# 0.1.1

* Bug fixes
  * use path option for releases

# 0.1.0

First release!


test/unimux_test.exs
defmodule UniMuxTest do
  use ExUnit.Case

  import Mock
  
  require Record
  Record.defrecordp(:context, Record.extract(:context, from_lib: "hello/include/hello.hrl"))

  setup do
    default_timeout = Application.get_env(:unimux, :default_timeout)
    on_exit fn -> 
      Application.put_env(:unimux, :routes, [])
      Application.put_env(:unimux, :default_timeout, default_timeout)
    end
  end

  test "handle request - not found" do
    state = []
    assert {:stop, :not_found, {:ok, :not_found}, state} == UniMux.handle_request(:context, "a.b.c", [], state)
  end

  test_with_mock "handle request - ok", :hello_client, [call: fn(name, _, timeout) -> {:ok, name, timeout} end] do
    state = []
    Application.put_env(:unimux, :routes, [{"a", 'http://127.0.0.1:8080', 10000}])
    assert UniMux.handle_request(:context, "a.b.c", [], state) == {:stop, :normal, {:ok, :hello_client_a, 10000}, state}  

    Application.put_env(:unimux, :routes, [{"a", 'http://127.0.0.1:8080', 10000}, {"a.b", "http://127.0.0.1:8081", 20000}])
    assert UniMux.handle_request(:context, "a.b.c", [], state) == {:stop, :normal, {:ok, :"hello_client_a.b", 20000}, state}
  end

  test "router" do
    id = 1
    assert {:ok, UniMux.name(), id} == UniMux.Router.route(context(session_id: id), :req, :uri)
  end

  test "server_timeout" do
    assert 10500 == Application.get_env(:hello, :server_timeout)
  end

  test_with_mock "default_timeout", :hello_client, [call: fn(name, _, timeout) -> {:ok, name, timeout} end] do
    state = []
    Application.put_env(:unimux, :routes, [{"a", 'http://127.0.0.1:8080', :undefined}, {"a.b", "http://127.0.0.1:8081", :undefined}])
    assert UniMux.handle_request(:context, "a.b.c", [], state) == {:stop, :normal, {:ok, :"hello_client_a.b", 10000}, state}

    Application.put_env(:unimux, :default_timeout, 20000)
    Application.put_env(:unimux, :routes, [{"a", 'http://127.0.0.1:8080', :undefined}, {"a.b", "http://127.0.0.1:8081", :undefined}])
    assert UniMux.handle_request(:context, "a.b.c", [], state) == {:stop, :normal, {:ok, :"hello_client_a.b", 20000}, state}
  end
end


test/hello_proxy_test.exs
defmodule HelloProxyTest do
  use ExUnit.Case

  test "the truth" do
    assert 1 + 1 == 2
  end
end


mix.exs
defmodule UniMux.Mixfile do
  use Mix.Project

  def project do
    [app: :unimux,
     version: "1.3.3",
     elixir: "~> 1.3.0",
     test_coverage: [tool: Coverex.Task, coveralls: true],
     build_embedded: Mix.env == :prod,
     start_permanent: Mix.env == :prod,
     deps: deps(Mix.env)]
  end

  def application do
    [applications: [:logger, :elixir, :metricman, :hello, :runtime_tools, :exrun | (if Mix.env == :release do [:lager_journald_backend] else [] end)],
     mod: {UniMux, []}]
  end

  @exclude_deps [:earmark, :ex_doc]
  defp deps(:release) do
    Code.eval_file("mix.lock")
    |> elem(0)
    |> Enum.filter_map(&(not (elem(&1, 0) in @exclude_deps)), fn({key, _}) -> {key, path: "deps/" <> "#{key}", override: true} end)
  end

  defp deps(_) do
    [{:lager, github: "basho/lager", tag: "3.2.2", override: true},
     {:goldrush, github: "DeadZen/goldrush", tag: "0.1.8", override: true},
     {:hello, github: "travelping/hello", branch: "master"},
     {:hackney, "~> 1.4.4", override: true},
     {:metricman, github: "xerions/metricman"},
     {:exlager, github: "xerions/exlager"},
     {:exrun, github: "liveforeverx/exrun"},
     {:conform_exrm, github: "bitwalker/conform_exrm"},
     # We use certain ref for conform because of bug with building of *.ez archive.
     # We should use version from hex when it will be avaliable
     {:conform, github: "bitwalker/conform", ref: "678dbfca6e7a94dd9d9214e6dbddf8b3ddb35dbf", override: true},
     {:coverex, github: "alfert/coverex", tag: "v1.4.3", only: :test},
     {:meck, github: "eproxus/meck", tag: "0.8.3", override: true},
     {:exometer_core, github: "Feuerlabs/exometer_core", override: true},
     {:erlware_commons, "~> 0.21.0", override: true},
     {:edown, github: "uwiger/edown", override: true},
     {:setup, github: "uwiger/setup", override: true},
     {:mock, github: "jjh42/mock", only: :test},
     {:relx, github: "erlware/relx", override: true}, #we need https://github.com/erlware/relx/issues/524
     {:exrm, "~> 1.0.8"}]
  end
end


lib/unimux.ex
defmodule UniMux do
  use Application

  def start(_type, _args) do
    import Supervisor.Spec, warn: false

    children = for client <- Application.get_env(:unimux, :routes, []) do
      {name, url, t} = client
      worker(Hello.Client, [{:local, namespace(name)}, url, {[{:recv_timeout, timeout(t)}], [], []}], id: namespace(name))
    end
    # Set hello server timeout to the maximum value of client's timeouts 
    # and add some  value for pretend errors
    timeouts = for {_, _, t} <- Application.get_env(:unimux, :routes, []), do: timeout(t)
    Application.put_env(:hello, :server_timeout, Enum.max(timeouts ++ [timeout(:defaut)]) + 500)
    listener_url = Application.get_env(:unimux, :listen, 'zmq-tcp://0.0.0.0')
    Hello.start_listener(listener_url, [], :hello_proto_jsonrpc, [], UniMux.Router)
    Hello.bind(listener_url, __MODULE__)
    opts = [strategy: :one_for_one, name: UniMux.Supervisor]
    Supervisor.start_link(children, opts)
  end

  def name(), do: Application.get_env(:unimux, :name, "proxy/server")
  def router_key(), do: ""
  def validation(), do: __MODULE__
  def request(_, m, p), do: {:ok, m, p}

  def init(_identifier, _), do: {:ok, Application.get_env(:unimux, :routes, [])}

  def handle_request(_context, method, args, state) do
    case resolve(method) do
      nil -> {:stop, :not_found, {:ok, :not_found}, state}
      {name, timeout} ->
        r = Hello.Client.call(name, {method, args, []}, timeout)
        {:stop, :normal, r, state}
    end
  end

  def handle_info(_context, _message, state) do
    {:noreply, state}
  end

  def terminate(_context, _reason, _state) do
    :ok
  end

  defp resolve(method) do
    names = for client <- Application.get_env(:unimux, :routes, []), do: {elem(client, 0), elem(client, 2)}
    splittedMethod = method |> :hello_lib.to_binary |> String.split(".", [:global])
    resolve(splittedMethod, names)
  end
  
  defp resolve([], _), do: nil
  defp resolve(_, []), do: nil
  defp resolve(method, names) do
    name = Enum.join(method, ".")
    case :lists.keyfind(name, 1, names) do
      {name, t} -> {namespace(name), timeout(t)}
      false -> List.delete_at(method, length(method) - 1) |>  resolve(names)
    end
  end

  defp timeout(t) when is_integer(t), do: t
  defp timeout(_), do: Application.get_env(:unimux, :default_timeout)

  defp namespace(name) do
    String.to_atom("hello_client_" <> name)
  end
end


defmodule UniMux.Router do
  require Record
  Record.defrecordp(:context, Record.extract(:context, from_lib: "hello/include/hello.hrl"))

  def route(context(session_id: id), _request, _uri) do
    {:ok, UniMux.name(), id}
  end
end


config/unimux.exs
[
  mappings: [
    "listen": [
      doc: """
      API endpoint in form of <protocol>://<host>[:<port>]

      Supported protocols are: zmq-tcp, zmq-tcp6, zmq-ipc, http

      It is possible to specify port as 0 or * to using only mdns registration
      """ ,
      to: "unimux.listen",
      datatype: :charlist,
      default: "zmq-tcp://127.0.0.1:20000"
    ],
    "route.*": [
      to: "unimux.routes",
      datatype: [:complex],
      default: []
    ],
    "route.*.pattern": [
      doc: "API prefix pattern",
      to: "unimux.routes",
      datatype: :binary,
      default: "APIPrefix"
    ],
    "route.*.target": [
      doc: "Route API endpoint in form of <protocol>://<host>[:<port>]",
      to: "unimux.routes",
      datatype: :string,
      default: "http://127.0.0.1:8080"
    ]
  ],
  translations: [
    "unimux.routes.*": fn _, {key, value_map}, acc ->
      [{value_map[:pattern], value_map[:target]} | acc]
    end,
    "listen": fn
      _, uri, acc ->
        try do
          :ex_uri.decode(uri)
          uri
        catch
          _, _ ->
            IO.puts("Unsupported URI format: #{uri}")
            exit(1)
        end
    end
  ]
]


lib/hello_proxy.ex
defmodule HelloProxy do
  # See http://elixir-lang.org/docs/stable/elixir/Application.html
  # for more information on OTP Applications
  use Application


  def start(_type, _args) do
    import Supervisor.Spec, warn: false

    children = for client <- Application.get_env(:hello_proxy, :routes, []) do
      {name, url} = client
      worker(:hello_client, [{:local, namespace(name)}, url, {[], [], []}], id: namespace(name))
    end
    listener_url = Application.get_env(:hello_proxy, :listener_url, 'zmq-tcp://0.0.0.0')
    :hello.start_service(__MODULE__, [])
    :hello.start_listener(listener_url, [], :hello_proto_jsonrpc, [], HelloProxy.Router)
    :hello.bind(listener_url, __MODULE__)
    # See http://elixir-lang.org/docs/stable/elixir/Supervisor.html
    # for other strategies and supported options
    opts = [strategy: :one_for_one, name: HelloProxy.Supervisor]
    Supervisor.start_link(children, opts)
  end

  # Hello service
  def name(), do: Application.get_env(:hello_proxy, :name, "proxy/server")
  def router_key(), do: ""
  def validation(), do: __MODULE__
  def request(_, m, p), do: {:ok, m, p}

  def init(_identifier, _), do: {:ok, Application.get_env(:hello_proxy, :routes, [])}

  def handle_request(_context, method, args, state) do
    splittedMethod = method |> :hello_lib.to_binary |> String.split(".", [:global])
    name = Enum.join(List.delete_at(splittedMethod, length(splittedMethod) - 1), ".")
    case Application.get_env(:hello_proxy, :routes) |> List.keyfind(name, 0) do
      nil -> {:stop, :normal, {:ok, :not_found}, state}
      _ ->
        r = namespace(name) |> :hello_client.call({method, args, []})
        {:stop, :normal, r, state}
    end
  end

  def handle_info(_context, _message, state) do
    {:noreply, state}
  end

  def terminate(_context, _reason, _state) do
    :ok
  end

  def namespace(name) do
    String.to_atom("hello_client_" <> name)
  end

  # for tests
  def client() do
    case :hello_client.start({:local, __MODULE__}, 'zmq-tcp://proxy/server', [], [], []) do
      {:ok, _} -> :timer.sleep(1000)
      _ -> []
    end
    :hello_client.call(__MODULE__, {"Test.try", [], []})
  end
end


defmodule HelloProxy.Router do
  require Record
  Record.defrecordp(:context, Record.extract(:context, from_lib: "hello/include/hello.hrl"))

  def route(context(session_id: id), _request, _uri) do
    {:ok, HelloProxy.name(), id}
  end
end


lib/hello_router.ex
defmodule HelloRouter do
  use Application

  def start(_type, _args) do
    import Supervisor.Spec, warn: false

    children = for client <- Application.get_env(:hello_router, :routes, []) do
      {name, url} = client
      worker(:hello_client, [{:local, namespace(name)}, url, {[], [], []}], id: namespace(name))
    end
    listener_url = Application.get_env(:hello_router, :listen, 'zmq-tcp://0.0.0.0')
    :hello.start_service(__MODULE__, [])
    :hello.start_listener(listener_url, [], :hello_proto_jsonrpc, [], HelloRouter.Router)
    :hello.bind(listener_url, __MODULE__)
    opts = [strategy: :one_for_one, name: HelloRouter.Supervisor]
    Supervisor.start_link(children, opts)
  end

  def name(), do: Application.get_env(:hello_router, :name, "proxy/server")
  def router_key(), do: ""
  def validation(), do: __MODULE__
  def request(_, m, p), do: {:ok, m, p}

  def init(_identifier, _), do: {:ok, Application.get_env(:hello_router, :routes, [])}

  def handle_request(_context, method, args, state) do
    splittedMethod = method |> :hello_lib.to_binary |> String.split(".", [:global])
    name = Enum.join(List.delete_at(splittedMethod, length(splittedMethod) - 1), ".")
    case Application.get_env(:hello_router, :routes) |> List.keyfind(name, 0) do
      nil -> {:stop, :normal, {:ok, :not_found}, state}
      _ ->
        r = namespace(name) |> :hello_client.call({method, args, []})
        {:stop, :normal, r, state}
    end
  end

  def handle_info(_context, _message, state) do
    {:noreply, state}
  end

  def terminate(_context, _reason, _state) do
    :ok
  end

  def namespace(name) do
    String.to_atom("hello_client_" <> name)
  end
end


defmodule HelloRouter.Router do
  require Record
  Record.defrecordp(:context, Record.extract(:context, from_lib: "hello/include/hello.hrl"))

  def route(context(session_id: id), _request, _uri) do
    {:ok, HelloRouter.name(), id}
  end
end

config/hello_router.schema.exs
[
  mappings: [
    "listen": [
      doc: """
      API endpoint in form of <protocol>://<host>[:<port>]

      Supported protocols are: zmq-tcp, zmq-tcp6, zmq-ipc, http

      It is possible to specify port as 0 or * to using only mdns registration
      """ ,
      to: "hello_router.listen",
      datatype: :charlist,
      default: "zmq-tcp://127.0.0.1:20000"
    ],
    "route.*": [
      to: "hello_router.routes",
      datatype: [:complex],
      default: []
    ],
    "route.*.pattern": [
      doc: "API prefix pattern",
      to: "hello_router.routes",
      datatype: :binary,
      default: "APIPrefix"
    ],
    "route.*.target": [
      doc: "Route API endpoint in form of <protocol>://<host>[:<port>]",
      to: "hello_router.routes",
      datatype: :string,
      default: "http://127.0.0.1:8080"
    ]
  ],
  translations: [
    "hello_router.routes.*": fn _, {key, value_map}, acc ->
      [{value_map[:pattern], value_map[:target]} | acc]
    end,
    "listen": fn
      _, uri, acc ->
        try do
          :ex_uri.decode(uri)
          uri
        catch
          _, _ ->
            IO.puts("Unsupported URI format: #{uri}")
            exit(1)
        end
    end
  ]
]


config/unimux.schema.exs
[
  extends: [
    :exlager
  ],
  mappings: [
    "listen": [
      doc: """
      API endpoint in form of <protocol>://<host>[:<port>]

      Supported protocols are: zmq-tcp, zmq-tcp6, zmq-ipc, http
      """ ,
      to: "unimux.listen",
      datatype: :charlist,
      default: "http://127.0.0.1:20000"
    ],
    "default_timeout": [
      doc: "Default timeout for all routes in milliseconds",
      to: "unimux.default_timeout",
      datatype: :integer,
      default: 10000
    ],
    "route.$id.pattern": [
      doc: "API prefix pattern",
      to: "unimux.$id.pattern",
      datatype: :binary,
      default: "APIPrefix"
    ],
    "route.$id.target": [
      doc: "Route API endpoint in form of <protocol>://<host>[:<port>]",
      to: "unimux.$id.target",
      datatype: :charlist,
      default: "http://127.0.0.1:8080",
    ],
    "route.$id.timeout": [
      doc: "Timeout for the route in milliseconds",
      to: "unimux.$id.timeout",
      datatype: :integer,
      default: :undefined
    ]
  ],
  transforms: [
    "unimux.routes": fn table ->
      patterns = Conform.Conf.get(table, "unimux.$id.pattern")
      targets = Conform.Conf.get(table, "unimux.$id.target")
      timeouts = Conform.Conf.get(table, "unimux.$id.timeout")
      Enum.map(patterns, fn({[_, id, _], pattern}) ->
        target = case Conform.Conf.get(table, "unimux." <> to_string(id) <> ".target") do
                   [{_, target}] ->
                     :ets.delete(table, ['unimux', id, 'target'])
                     target
                   _ ->
                     "http://127.0.0.1:8080"
                 end
        timeout = case Conform.Conf.get(table, "unimux." <> to_string(id) <> ".timeout") do
                   [{_, timeout}] ->
                      :ets.delete(table, ['unimux', id, 'timeout'])
                      timeout
                   _ ->
                     :undefined
                  end
        :ets.delete(table, ['unimux', id, 'pattern'])
        :ets.match_delete(table, {['unimux', '$id' | :_], :_})
        {pattern, target, timeout}
      end)
    end,
    "unimux.listen": fn table ->
      [{_, uri}] = Conform.Conf.get(table, "unimux.listen")
      case :ex_uri.decode(uri) do
          {:error, :invalid_uri} ->
            IO.puts("#{IO.ANSI.red}Unsupported URI format for 'listen' (API endpoint) option: #{uri}#{IO.ANSI.reset}")
            exit("Unsupported URI format for 'listen' (API endpoint) option: #{uri}")
          _ ->
            uri
        end
    end
  ]
]
